File: hawkey/tests/python/tests/test_reldep.py
#
# Copyright (C) 2012-2013 Red Hat, Inc.
#
# Licensed under the GNU Lesser General Public License Version 2.1
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#

from __future__ import absolute_import
from . import base

import hawkey

class Reldep(base.TestCase):
    def setUp(self):
        self.sack = base.TestSack(repo_dir=self.repo_dir)
        self.sack.load_system_repo()

    def test_basic(self):
        flying = base.by_name(self.sack, "flying")
        requires = flying.requires
        self.assertLength(requires, 1)
        reldep = requires[0]
        self.assertEqual(str(reldep), "P-lib >= 3")

    def test_custom_creation(self):
        reldep_str = "P-lib >= 3"
        reldep = hawkey.Reldep(self.sack, reldep_str)
        self.assertEqual(reldep_str, str(reldep))

    def test_custom_creation_fail(self):
        reldep_str = "lane = 4"
        self.assertRaises(hawkey.ValueException, hawkey.Reldep, self.sack,
                          reldep_str)
        reldep_str = "P-lib >="
        self.assertRaises(hawkey.ValueException, hawkey.Reldep, self.sack,
                          reldep_str)

    def test_custom_querying(self):
        reldep = hawkey.Reldep(self.sack, "P-lib = 3-3")
        q = hawkey.Query(self.sack).filter(provides=reldep)
        self.assertLength(q, 1)
        reldep = hawkey.Reldep(self.sack, "P-lib >= 3")
        q = hawkey.Query(self.sack).filter(provides=reldep)
        self.assertLength(q, 1)
        reldep = hawkey.Reldep(self.sack, "P-lib < 3-3")
        q = hawkey.Query(self.sack).filter(provides=reldep)
        self.assertLength(q, 0)

    def test_query_name_only(self):
        reldep_str = "P-lib"
        reldep = hawkey.Reldep(self.sack, reldep_str)
        self.assertEqual(reldep_str, str(reldep))
        q = hawkey.Query(self.sack).filter(provides=reldep)
        self.assertLength(q, 1)
        self.assertEqual(str(q[0]), "penny-lib-4-1.x86_64")

    def test_not_crash(self):
        self.assertRaises(ValueError, hawkey.Reldep)

    def test_non_num_version(self):
        reldep_str = 'foo >= 1.0-1.fc20'
        with self.assertRaises(hawkey.ValueException) as ctx:
            hawkey.Reldep(self.sack, reldep_str)
        self.assertEqual(ctx.exception.args[0], "No such reldep: %s" % reldep_str)

File: hawkey/tests/test_reldep.c
/*
 * Copyright (C) 2012-2013 Red Hat, Inc.
 *
 * Licensed under the GNU Lesser General Public License Version 2.1
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

// hawkey
#include "src/package.h"
#include "src/reldep.h"
#include "src/sack.h"
#include "fixtures.h"
#include "test_suites.h"
#include "testsys.h"

START_TEST(test_reldeplist_add)
{
    HySack sack = test_globals.sack;
    HyPackage flying = by_name_repo(sack, "fool", "updates");
    HyReldepList reldeplist = hy_reldeplist_create(sack);
    HyReldepList obsoletes = hy_package_get_obsoletes(flying);

    const int count = hy_reldeplist_count(obsoletes);
    fail_unless(count == 2);
    for (int i = 0; i < count; ++i) {
	HyReldep reldep = hy_reldeplist_get_clone(obsoletes, i);
	hy_reldeplist_add(reldeplist, reldep);
	hy_reldep_free(reldep);
    }

    hy_package_free(flying);
    fail_unless(hy_reldeplist_count(reldeplist) == 2);
    hy_reldeplist_free(obsoletes);
    hy_reldeplist_free(reldeplist);
}
END_TEST

Suite *
reldep_suite(void)
{
    Suite *s = suite_create("Reldep");
    TCase *tc = tcase_create("Core");
    tcase_add_unchecked_fixture(tc, fixture_with_updates, teardown);
    tcase_add_test(tc, test_reldeplist_add);
    suite_add_tcase(s, tc);

    return s;
}
